File: 3756-concatenate-non-zero-digits-and-multiply-by-sum-ii/3756-concatenate-non-zero-digits-and-multiply-by-sum-ii.ts
const MOD = 1000000007;
const MAX = 100001;

const pow = new Int32Array(MAX);
pow[0] = 1;

for (let i = 1; i < MAX; i++)
    pow[i] = (pow[i - 1] * 10) % MOD;

const sumAndMultiply = (s, queries) => {
    const n = s.length;
    const A = new Int32Array(n + 1);
    const B = new Int32Array(n + 1);
    const len = new Int32Array(n + 1);

    for (let i = 0; i < n; i++) {
        const d = Number(s[i]);
        A[i + 1] = A[i] + d;
        B[i + 1] = d ? (B[i] * 10 + d) % MOD : B[i];
        len[i + 1] = len[i] + (d ? 1 : 0);
    }

    const res = [];

    for (let [l, r] of queries) {
        r++;

        const ln = len[r] - len[l];
        const sub = Number((BigInt(B[l]) * BigInt(pow[ln])) % BigInt(MOD));
        let x = B[r] - sub;

        x = ((x % MOD) + MOD) % MOD;

        res.push((x * (A[r] - A[l])) % MOD);
    }

    return res;
};

File: 3756-concatenate-non-zero-digits-and-multiply-by-sum-ii/3756-concatenate-non-zero-digits-and-multiply-by-sum-ii.rs
impl Solution {
    const MOD: i64 = 1_000_000_007;
    const MAX: usize = 100_001;
    
    pub fn sum_and_multiply(s: String, queries: Vec<Vec<i32>>) -> Vec<i32> {
        let s = s.as_bytes();
        let n = s.len();
        
        let mut pow10 = vec![0i64; Self::MAX];
        pow10[0] = 1;
        for i in 1..Self::MAX {
            pow10[i] = (pow10[i - 1] * 10) % Self::MOD;
        }
        
        let mut A = vec![0i64; n + 1];
        let mut B = vec![0i64; n + 1];
        let mut len = vec![0i64; n + 1];
        
        for i in 0..n {
            let d = (s[i] - b'0') as i64;
            
            A[i + 1] = A[i] + d;
            
            if d != 0 {
                B[i + 1] = (B[i] * 10 + d) % Self::MOD;
                len[i + 1] = len[i] + 1;
            } else {
                B[i + 1] = B[i];
                len[i + 1] = len[i];
            }
        }
        
        let mut results = Vec::with_capacity(queries.len());
        
        for query in &queries {
            let l = query[0] as usize;
            let r = query[1] as usize + 1;
            
            let ln = (len[r] - len[l]) as usize;
            let sub = (B[l] * pow10[ln]) % Self::MOD;
            let mut x = (B[r] - sub) % Self::MOD;
            if x < 0 { x += Self::MOD; }
            
            results.push(((x * (A[r] - A[l])) % Self::MOD) as i32);
        }
        
        results
    }
}